### PR TITLE
NEP: New NEP for identifying and coercing duck arrays

### PR DESCRIPTION
**Pull request by njsmith**

### CODE REVIEW

File: doc/neps/nep-0016-abstract-array.rst [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+To start with, we should declare ``__array_ufunc__`` to be an abstract
+method, and add the ``NDArrayOperatorsMixin`` methods as mixin
+methods.
+
+Declaring ``astype`` as an ``@abstractmethod`` probably makes sense as
+well, since it's used by ``asabstractarray``. We might also want to go
+ahead and add some basic attributes like ``ndim``, ``shape``,
+``dtype``.
+
+Adding new abstract methods will be a bit trick, because ABCs enforce
```
Comment (eric-wieser) [inline doc/neps/nep-0016-abstract-array.rst]: typo: `tricky`

File: doc/neps/nep-0016-abstract-array.rst [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+``asarray`` should be replaced by this, and in the future it's what
+everyone should be reaching for by default unless they have a specific
+reason to use ``asarray`` instead. This suggests that its name really
+should be *shorter* and *more memorable* than ``asarray``... which
+is difficult. I've used ``asabstractarray`` in this draft, but I'm not
+really happy with it, because it's too long and people are unlikely to
+start using it by habit without endless exhortations.
+
+One option would be to actually change ``asarray``\'s semantics so
+that *it* passes through ``AbstractArray`` objects unchanged. But I'm
```
Comment (eric-wieser) [inline doc/neps/nep-0016-abstract-array.rst]: Perhaps just change `asanyarray`?

File: doc/neps/nep-0016-abstract-array.rst [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+object dtype, or raise an error, or similar. Probably none of these
+outcomes are actually desireable in most cases, so maybe making it a
+segfault instead would be OK? But it's dangerous given that we don't
+know how common such code is. OTOH, if we were starting from scratch
+then this would probably be the ideal solution.
+
+We can't use ``asanyarray`` or ``array``, since those are already
+taken.
+
+Any other ideas? ``np.cast``, ``np.coerce``?
```
Comment (eric-wieser) [inline doc/neps/nep-0016-abstract-array.rst]: `np.cast[np.float16](x)` is already a thing

File: doc/neps/nep-0016-abstract-array.rst [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+
+
+Implementation
+--------------
+
+1. Rename ``NDArrayOperatorsMixin`` to ``AbstractArray`` (leaving
+   behind an alias for backwards compatibility) and make it an ABC.
```
Comment (eric-wieser) [inline doc/neps/nep-0016-abstract-array.rst]: Why not keep them separate, and have `AbstractArray` subclass the mixin? What about users who just want to use ufuncs, but no other part of the array interface?

**Comment by charris** [comment]
Needs rebase to get rid of circleci error.

**Comment by shoyer** [comment]
Maybe we can merge this in https://github.com/numpy/numpy/pull/12174 instead?

**Comment by mattip** [comment]
Closing in favor of #12174

**Pull request closed without merging**